### PR TITLE
pufferpanel: init at 2.2.0

### PR DESCRIPTION
**Pull request by ckiee**
###### Motivation for this change
Friend wanted PufferPanel packaged.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by ronthecookie** [comment]
Result of `nixpkgs-review pr 128243` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>pufferpanel</li>
  </ul>
</details>

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 128243` at 7ee44ca4 run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built successfully:</summary>
  <ul>
    <li>pufferpanel</li>
  </ul>
</details>
<details>
  <summary>3 suggestions:</summary>
  <ul>
    <li>warning: unused-argument

Unused argument: `stdenv`.
Near pkgs/servers/pufferpanel/default.nix:7:3:
```
  |
7 | , stdenv
  |   ^
```
</li>
    <li>warning: unused-argument

Unused argument: `pkgs`.
Near pkgs/servers/pufferpanel/default.nix:6:3:
```
  |
6 | , pkgs
  |   ^
```
</li>
    <li>warning: unused-argument

Unused argument: `nodejs`.
Near pkgs/servers/pufferpanel/default.nix:5:3:
```
  |
5 | , nodejs
  |   ^
```
</li>
  </ul>
</details>

**Comment by ronthecookie** [comment]
Just discovered it requires java to be in the PATH alongside other things, marking as draft for now.

File: pkgs/servers/pufferpanel/default.nix [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+  vendorSha256 = "061l1sy0z3kd7rc2blqh333gy66nbadfxy9hyxgq07dszds4byys";
+
+  postFixup = ''
+    mkdir $out/share
+    cp -r ${src}/assets/email $out/share/templates
+    cp -r ${frontend} $out/share/www
+
+    # Wrap the binary with the path to the external files.
+    mv $out/bin/cmd $out/bin/pufferpanel
+    wrapProgram "$out/bin/pufferpanel" --set PUFFER_PANEL_EMAIL_TEMPLATES $out/share/templates/emails.json --set GIN_MODE release --set PUFFER_PANEL_WEB_FILES $out/share/www
```
Comment (SuperSandro2000) [inline pkgs/servers/pufferpanel/default.nix:44]:
```suggestion
    wrapProgram "$out/bin/pufferpanel" \
      --set PUFFER_PANEL_EMAIL_TEMPLATES $out/share/templates/emails.json \
      --set GIN_MODE release --set PUFFER_PANEL_WEB_FILES $out/share/www
```

**Comment by ronthecookie** [comment]
Further investigation revealed the required binaries change at runtime (and it's not feasible to include every single possible dependency) so this package so I've adapted the package for something like:
```nix
pkgs.pufferpanel.override {
    pathDeps = [ pkgs.openjdk16 ];
}
```

Oh also, thanks for the review as usual Sandro (:

File: pkgs/servers/pufferpanel/default.nix [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  postFixup = ''
+    mkdir -p $out/share/pufferpanel
+    cp -r ${src}/assets/email $out/share/pufferpanel/templates
+    cp -r ${frontend} $out/share/pufferpanel/www
+
+    # Wrap the binary with the path to the external files.
+    mv $out/bin/cmd $out/bin/pufferpanel
+    wrapProgram "$out/bin/pufferpanel" \
+      --set PUFFER_PANEL_EMAIL_TEMPLATES $out/share/pufferpanel/templates/emails.json \
+      --set GIN_MODE release --set PUFFER_PANEL_WEB_FILES $out/share/pufferpanel/www \
```
Comment (SuperSandro2000) [inline pkgs/servers/pufferpanel/default.nix:49]:
```suggestion
      --set GIN_MODE release \
      --set PUFFER_PANEL_WEB_FILES $out/share/pufferpanel/www \
```

File: pkgs/servers/pufferpanel/default.nix [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+  # frontend.
+  # Getting the frontend to build in the Nix environment fails even
+  # with all the proper node_modules populated. To work around this,
+  # we just download the built frontend and package that.
+  frontend = fetchzip
+    {
+      url = "https://github.com/PufferPanel/PufferPanel/releases/download/v${version}/pufferpanel_${version}_linux_arm64.zip";
+      sha256 = "0phbf4asr0dns7if84crx05kfgr44yaxrbsbihdywbhh2mb16052";
+      stripRoot = false;
+    } + "/www";
```
Comment (SuperSandro2000) [inline pkgs/servers/pufferpanel/default.nix:34]:
```suggestion
  frontend = fetchzip {
    url = "https://github.com/PufferPanel/PufferPanel/releases/download/v${version}/pufferpanel_${version}_linux_arm64.zip";
    sha256 = "0phbf4asr0dns7if84crx05kfgr44yaxrbsbihdywbhh2mb16052";
    stripRoot = false;
  } + "/www";
```

**Comment by SuperSandro2000** [comment]
@ofborg eval

**Pull request merged**